From 685e20421c7ad7b431bd6f15ad058b1129d332dd Mon Sep 17 00:00:00 2001
From: Niall Pemberton <niallp@apache.org>
Date: Fri, 29 Jan 2010 19:18:36 +0000
Subject: [PATCH] Port LANG-432 to 2.x branch - Applying the final part of
 Benjamin Bentmann's patch to LANG-432, improving handling of case-insensitive
 Strings

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/lang/branches/LANG_2_X@904598 13f79535-47bb-0310-9956-ffa450edef68
---
 .../org/apache/commons/lang/StringUtils.java  | 13 +++++--
 .../lang/StringUtilsEqualsIndexOfTest.java    | 38 ++++++++++++++++++-
 2 files changed, 47 insertions(+), 4 deletions(-)

diff --git a/src/main/java/org/apache/commons/lang/StringUtils.java b/src/main/java/org/apache/commons/lang/StringUtils.java
index 4644b04e8df..7a0cdecaf6a 100644
--- a/src/main/java/org/apache/commons/lang/StringUtils.java
+++ b/src/main/java/org/apache/commons/lang/StringUtils.java
@@ -1254,8 +1254,8 @@ public static boolean contains(String str, String searchStr) {
 
     /**
      * <p>Checks if String contains a search String irrespective of case,
-     * handling <code>null</code>. This method uses
-     * {@link #contains(String, String)}.</p>
+     * handling <code>null</code>. Case-insensitivity is defined as by
+     * {@link String#equalsIgnoreCase(String)}.
      *
      * <p>A <code>null</code> String will return <code>false</code>.</p>
      *
@@ -1279,7 +1279,14 @@ public static boolean containsIgnoreCase(String str, String searchStr) {
         if (str == null || searchStr == null) {
             return false;
         }
-        return contains(str.toUpperCase(), searchStr.toUpperCase());
+        int len = searchStr.length();
+        int max = str.length() - len;
+        for (int i = 0; i <= max; i++) {
+            if (str.regionMatches(true, i, searchStr, 0, len)) {
+                return true;
+            }
+        }
+        return false;
     }
 
     // IndexOfAny chars
diff --git a/src/test/java/org/apache/commons/lang/StringUtilsEqualsIndexOfTest.java b/src/test/java/org/apache/commons/lang/StringUtilsEqualsIndexOfTest.java
index b2204f30474..b074c3d0901 100644
--- a/src/test/java/org/apache/commons/lang/StringUtilsEqualsIndexOfTest.java
+++ b/src/test/java/org/apache/commons/lang/StringUtilsEqualsIndexOfTest.java
@@ -16,6 +16,8 @@
  */
 package org.apache.commons.lang;
 
+import java.util.Locale;
+
 import junit.framework.Test;
 import junit.framework.TestCase;
 import junit.framework.TestSuite;
@@ -395,7 +397,41 @@ public void testContainsIgnoreCase_StringString() {
         assertTrue(StringUtils.containsIgnoreCase("xabcz", "ABC"));
     }
 
-    //-----------------------------------------------------------------------
+    public void testContainsIgnoreCase_LocaleIndependence() {
+        Locale orig = Locale.getDefault();
+
+        Locale[] locales = { Locale.ENGLISH, new Locale("tr", ""), Locale.getDefault() };
+
+        String[][] tdata = { 
+            { "i", "I" },
+            { "I", "i" },
+            { "\u03C2", "\u03C3" },
+            { "\u03A3", "\u03C2" },
+            { "\u03A3", "\u03C3" },
+        };
+
+        String[][] fdata = { 
+            { "\u00DF", "SS" },
+        };
+
+        try {
+            for (int i = 0; i < locales.length; i++) {
+                Locale.setDefault(locales[i]);
+                for (int j = 0; j < tdata.length; j++) {
+                    assertTrue(Locale.getDefault() + ": " + j + " " + tdata[j][0] + " " + tdata[j][1], StringUtils
+                            .containsIgnoreCase(tdata[j][0], tdata[j][1]));
+                }
+                for (int j = 0; j < fdata.length; j++) {
+                    assertFalse(Locale.getDefault() + ": " + j + " " + fdata[j][0] + " " + fdata[j][1], StringUtils
+                            .containsIgnoreCase(fdata[j][0], fdata[j][1]));
+                }
+            }
+        } finally {
+            Locale.setDefault(orig);
+        }
+    }
+
+    // -----------------------------------------------------------------------
     public void testIndexOfAny_StringStringarray() {
         assertEquals(-1, StringUtils.indexOfAny(null, (String[]) null));
         assertEquals(-1, StringUtils.indexOfAny(null, FOOBAR_SUB_ARRAY));
